projects/sampling/executor.py

```diff
@@ -0,0 +1,185 @@
+        input_dir = uri
+        output_dir = artifact_utils.get_split_uri([output_artifact], split)
+        os.mkdir(output_dir)
+        for filename in fileio.listdir(input_dir):
+          input_uri = os.path.join(input_dir, filename)
+          output_uri = os.path.join(output_dir, filename)
+          io_utils.copy_file(src=input_uri, dst=output_uri, overwrite=True)
+
+  def undersample(self, uri, label, shards, keep_classes, output_dir):
+    """Function that actually undersamples the given split.
+
+    Args:
+      uri: The input uri for the specific split of the input example artifact.
+      label: The name of the column containing class names to undersample by.
+      shards: The number of files that each undersampled split should
+      contain. Default 0 is Beam's tfrecordio function's default.
+      keep_classes: A list determining which classes that we should
+      not undersample. Defaults to None.
+      output_dir: The output directory for the split of the output artifact.
+
+    Returns:
+      None
+    """
+
+    def generate_elements(x):
+      class_label = None
+      parsed = tf.train.Example.FromString(x.numpy())
+      if parsed.features.feature[label].int64_list.value:
+        val = parsed.features.feature[label].int64_list.value
+        if len(val) > 0:
+          class_label = val[0]
+      else:
+        val = parsed.features.feature[label].bytes_list.value
+        if len(val) > 0:
+          class_label = val[0].decode()
+      return (class_label, parsed)
+
+    def sample(key, value, side=0):
+      for item in random.sample(value, side):
+        yield item
+
+    def filter_null(item, keep_null=False, null_vals=None, pr=False):
+      if pr:
+        print(item)
+      keep = (not item[0]) or item[0] == 0  # Note that 0 is included in this filter!
+      if null_vals:
+        keep = keep or str(item[0]) in null_vals
+      keep ^= not keep_null  # null is True if we want to keep nulls
+      if keep:
+        return item[1]
+  
+    files = [os.path.join(uri, name) for name in os.listdir(uri)]
+    dataset = tf.data.TFRecordDataset(files, compression_type="GZIP")
+
+    with beam.Pipeline() as p:
+      # Take the input TFRecordDataset and extract the class label that we want.
+      # Output format is a K-V PCollection: {class_label: TFRecord in string format}
+      data = (
+        p
+        | "DatasetToPCollection" >> beam.Create(dataset)
+        | "MapToLabel" >> beam.Map(generate_elements)
+      )
+
+      # Finds the minimum frequency of all classes in the input label.
+      # Output is a singleton PCollection with the minimum # of examples.
+      val = beam.pvalue.AsSingleton(
+        (
+          data
+          | "CountPerKey" >> beam.combiners.Count.PerKey()
+          | "FilterNull" >> beam.Filter(lambda x: filter_null(x, null_vals=keep_classes))
+          | "Values" >> beam.Values()
+          | "FindMinimum" >> beam.CombineGlobally(lambda elements: min(elements or [-1]))
```

Comment:
I'm unsure of what this `[-1]` is supposed. to help with. When I tried to run this the first time in my pipeline this was being returned. Since `random.sample()` will error if `side=-1` this was causing issues. I'm not too versed with beam pipelines so this confused me. I changed this to `[0]` to fix the issue, but I'm not sure if I might have created another issue in the process.

Reply:
The `[-1]` is supposed to provide a fallback just in case no non-null elements are present after the `filter_null` call. This causes an error in `random.sample()`, which would signal that something went wrong during pipeline processing. The fact that you reach `-1` in the first place is worrying - what kinds of data did you input into the pipeline?

Reply:
I noticed some bugs within the Beam pipeline, so I fixed them - try running your code again to see if you run into the `[-1]` value again.